### PR TITLE
Implement Blob module

### LINKED ISSUE #3178: Add BlobManager support
Better Blob support was added in react-native 0.60, but requires a native module to work, which isn't implemented in react-native-windows.

Blobs allow better large file network calls using streams, rather than everything ending up in a single string.

Currently the networking JS call throws internal exceptions attempting to use blobs, which makes it more annoying to have break on exception turned on when debugging JS code.

### LINKED ISSUE #6093: Avoid multiple allocations in BlobModule
See https://github.com/microsoft/react-native-windows/pull/6079/files/359f9c92e0143b89d3d40efa1a2ad54c06871bf0#r494023291

Type `std::vector<std::uint8_t>` is used to both store and expose binary data by `WebSocketModule`.
When the blob type is used, and a segment projection is required (see `IWebSocketContentHandler::ResolveMessage`), a new vector object is created, copying the immutable data over.

Find a way to create this "slice" vector without copying, but also without modifying interface signatures in `IWebSocketResource`, or determine the proper storage/transfer type for binary data.
This should accommodate both `blob` and `arrayBuffer` binary types.

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description
Implements the Blob data type for both HTTP and WebSocket modules.

### New APIs
- `Microsoft::React::BlobModule`
- `Microsoft::React::IBlobPersistor`
- `Microsoft::React::IHttpModuleProxy`
- `Microsoft::React::IRequestBodyHandler`
- `Microsoft::React::IResponseHandler`
- `Microsoft::React::IUriHandler`
- `Microsoft::React::IWebSocketModuleContentHandler`
- `Microsoft::React::IWebSocketModuleProxy`

### Updated APIs
- `Microsoft::React::Networking::IHttpResource`
  - SendRequest
    Use `folly::dynamic` instead of `BodyData` for request body
- `SetOnRequest` renamed to `SetOnRequestSuccess` for clarity
- Added `SetOnBlobData` to handle complex/structured response content.

### Updated ABIs
- Removed `WebSocketModule` constructor.

### Type of Change
- New feature (non-breaking change which adds functionality)

### Why
Blob object types (i.e. File, Image) are part of the expected React Native functionality.
This feature exists in iOS, Android and macOS. It is missing on Windows.

Supersedes #6079
Resolves #3178
Fixes #6093

## Testing
- Added optional RNTester integration test WebSocketBlobTest`.

## Pre-merge checklist
- [x] #6093

## Follow-up work
- [ ] Implement BlobCollector (see https://github.com/facebook/react-native/pull/24745).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9352)

### CODE REVIEW

**Comment by rozele** [comment]
Cool! What's the plan for integrating this with the ImageViewManager / ImageModule / NetworkingModule?

**Comment by rozele** [comment]
Also, we should consider how we can leverage this module outside the ABI from external modules and view managers. For example, I'd expect a clipboard module may want to use the BlobModule when copying and pasting files / images / etc.

**Comment by JunielKatarn** [comment]
> Cool! What's the plan for integrating this with the ImageViewManager / ImageModule / NetworkingModule?

There are no specific plans for `ImageViewManager` or `ImageModule`. If they are directly dependent on Blob/BlobManager in iOS/Android, I'll make the necessary changes on a subsequent PR.

About `NetworkingModule`, the same, but it should go into this PR before merging.

> Also, we should consider how we can leverage this module outside the ABI from external modules and view managers. For example, I'd expect a clipboard module may want to use the BlobModule when copying and pasting files / images / etc.

I'm not following.
ABI should not matter, for the module is meant to be consumed/projected via JavaScript.

**Comment by asklar** [comment]
> > Cool! What's the plan for integrating this with the ImageViewManager / ImageModule / NetworkingModule?
> 
> There are no specific plans for `ImageViewManager` or `ImageModule`. If they are directly dependent on Blob/BlobManager in iOS/Android, I'll make the necessary changes on a subsequent PR.
> 
> About `NetworkingModule`, the same, but it should go into this PR before merging.
> 
> > Also, we should consider how we can leverage this module outside the ABI from external modules and view managers. For example, I'd expect a clipboard module may want to use the BlobModule when copying and pasting files / images / etc.
> 
> I'm not following. ABI should not matter, for the module is meant to be consumed/projected via JavaScript.

I think @rozele is referring to how an external native module (e.g. clipboard) might want to use blob storage to copy/paste data into/out of blobs. As it is, native modules won't have access to the module from native code since they are not exposed through the WinRT ABI.

**Comment by JunielKatarn** [comment]
> I think @rozele is referring to how an external native module (e.g. clipboard) might want to use blob storage to copy/paste data into/out of blobs. As it is, native modules won't have access to the module from native code since they are not exposed through the WinRT ABI.

Then I think we could try and implement the corresponding TurboModule for MSRN, but also not sure whether this would go into this PR.

**Comment by asklar** [comment]
> > I think @rozele is referring to how an external native module (e.g. clipboard) might want to use blob storage to copy/paste data into/out of blobs. As it is, native modules won't have access to the module from native code since they are not exposed through the WinRT ABI.
> 
> Then I think we could try and implement the corresponding TurboModule for MSRN, but also not sure whether this would go into this PR.

turbomodules are not going to be abi safe either, so there would need to be a way to access blob module in an abi safe way (i.e. through the WinRT API surface)

**Comment by acoates-ms** [comment]
A few things.  It looks like this wont actually hook up the new module to RNW.  Shouldn't we add it as a default module, the same as the websocket module?  (InstanceImpl::GetDefaultNativeModules)
That way both win32 and the public RNW will get the new functionality, and we wont be adding to the debt of exported non-ABI safe functions.

It pains me a little to be adding yet another cxxmodule, instead of writing this as a TurboModule which would verify the interface against the spec and help us on our way to reduce / remove the usage of folly.  Unfortunately until we get office onto the new ABI host interface (I'm working on it), office isn't able to run the newer TurboModules, so if we need blob support in Office within the next couple of months then we'd have to stick with the cxxmodule.  However, it will need to be created and added to the instance internally, like websockets.  Because when using the ABI host, cxxmodules are not supported.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [inline vnext/Shared/Modules/BlobModule.cpp:105]: Before merging:

Allow invoking this functionality directly on the native thread to avoid an unnecessary round trip.

Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:104]: Update.

Will communicate with the `WebSocketModule` via interface `IWebSocketModuleProxy`.

File: vnext/Shared/Modules/BlobModule.h
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.h:27]: >mutex [](http://example.com/codeflow?start=7&length=5)

Two mutexes in the same class is too much. It usually causes a deadlock. It is better to keep not more than one.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:45]: This code is not thread safe. In order to access any static variable from different threads we must have a static mutex. 
Why do we have multiple threads? 
There can be multiple RN instances and each will have its own thread. The s_contentHandler is going to be shared across multiple instances.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:33]: >IWebSocketModuleContentHandler [](http://example.com/codeflow?start=27&length=30)

Why not to keep BlobWebSocketModuleContentHandler here?

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:188]: >socketID [](http://example.com/codeflow?start=22&length=8)

We effectively search twice. It is better to erase the iterator received from the find method.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:197]: > return winrt::array_view<uint8_t>{data} [](http://example.com/codeflow?start=1&length=40)

It is not safe to return effectively a pointer to a data that can be changed in a different thread. While it may be inefficient we must return a copy. Though if the blobs are immutable, then we could keep shared_ptr to them in m_blobs and return just such shared_ptr.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:209]: >m_blobs [](http://example.com/codeflow?start=2&length=7)

If we go with shared_ptr keeping an immutable blob, then code here must create a new shared_ptr to the new blob. All readers of the previous shared_ptr will safely read the old copy.

File: vnext/Shared/Modules/WebSocketModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/WebSocketModule.cpp:154]: >s_sharedState [](http://example.com/codeflow?start=2&length=13)

In the system that has multiple instances it must not work. Two instances will race for this variable and it sounds wrong to me.

**Comment by JunielKatarn** [comment]
> Cool! What's the plan for integrating this with the ImageViewManager / ImageModule / NetworkingModule? Looks like Android uses a registration mechanism in the Android manifest to map `content://` URIs to the BlobProvider implementation. I don't think we have anything like this for WinUI, would this require custom handling in each module that works with blobs?

Image modules are out of the scope of this PR (if there is concrete missing functionality, kindly log an issue for triaging).

HTTP module integration is almost finished.
It will follow a similar design as ReactAndroid.
It will be delivered in this PR.

File: vnext/src/Libraries/Network/RCTNetworkingWinShared.js
Comment (JunielKatarn) [inline vnext/src/Libraries/Network/RCTNetworkingWinShared.js:64]: As seen in RCTNetworking.androkd.js

**Comment by JunielKatarn** [comment]
> Also, we should consider how we can leverage this module outside the ABI from external modules and view managers. For example, I'd expect a clipboard module may want to use the BlobModule when copying and pasting files / images / etc.

The current interfaces used to implement Blob interaction are internal-only and non-ABI-safe.
This is a possible improvement, but not in the scope of this first PR.

**Comment by JunielKatarn** [comment]
Blocked by #9939 (unrelated).

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:282]: >winrt::array_view<uint8_t>{data}; [](http://example.com/codeflow?start=9&length=33)

This should be sliced to (offset, size)

File: vnext/Shared/Networking/IHttpResource.h
Comment (mganandraj) [inline vnext/Shared/Networking/IHttpResource.h:95]: >SetOnBlobData [](http://example.com/codeflow?start=15&length=13)

It doesn't have to be blobs, right .. The data depends on whichever urihandler handles is picked, Blob handler could be just one of them ..

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:348]: >BlobModuleUriHandler [](http://example.com/codeflow?start=15&length=20)

Looks like you have implemented assuming that the uri is a local filesystem path .. Is there a way to create custom url schemes on windows, akin to
iOS : [https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app)
Android : [https://developer.android.com/guide/topics/providers/content-providers](https://developer.android.com/guide/topics/providers/content-providers)

If not, the URI based blob resolution may not make sense on Windows .. One solution will be to use a temporary file based blob persister and use the file path as the URI though ..

Comment (mganandraj) [reply vnext/Shared/Modules/BlobModule.cpp:348]: This would be a requirement for integrating blobs to ImageViewManager for instance.

File: vnext/Shared/Networking/WinRTHttpResource.cpp
Comment (mganandraj) [inline vnext/Shared/Networking/WinRTHttpResource.cpp:198]: Unless we implement custom URIs, this URI handler block doesn't make sense .. HTTP integration won't need it ..

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:134]: >BlobModule [](http://example.com/codeflow?start=12&length=10)

Julio, I don't understand how can someone use the Http/Fetch with blobs without implementing FileReaderModule !! Or am i missing something here ?

Comment (mganandraj) [reply vnext/Shared/Modules/BlobModule.cpp:134]: [https://github.com/facebook/react-native/blob/1b31d6bb582768ccbe92d3c1a9e43354a8c531e5/Libraries/Blob/FileReader.js#L122](https://github.com/facebook/react-native/blob/1b31d6bb582768ccbe92d3c1a9e43354a8c531e5/Libraries/Blob/FileReader.js#L122)

File: vnext/Shared/OInstance.cpp
Comment (mganandraj) [inline vnext/Shared/OInstance.cpp:551]: >transitionalProps [](http://example.com/codeflow?start=20&length=17)

Do we want to share the property bag across modules ? Would this lambda capture end up copying the property bag ?

File: vnext/Shared/Modules/BlobModule.cpp
Comment (vmoroz) [inline vnext/Shared/Modules/BlobModule.cpp:280]: >at [](http://example.com/codeflow?start=23&length=2)

What if it is not found? It will cause the app crash, right?

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:146]: Should this be in addWebSocketHandler ?

File: vnext/Shared/Networking/IHttpResource.h
Comment (JunielKatarn) [reply vnext/Shared/Networking/IHttpResource.h:95]: Agreed. This is not the best name. I'll update.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:348]: I did have that assumption.
I couldn't infer otherwise from the Android code base.

Following up.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:134]: Will discuss offline.
This is likely a misunderstanding of the Android code.

File: vnext/Shared/OInstance.cpp
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:551]: Yes, and yes.

On the deep dive describing this architecture, we decided to use an exclusive property bag to share objects between CxxModules.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:280]: True. Solved by resorting to the iterator.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:146]: Not really.
See Android's implementation:
```Java
  @Override
  public void addWebSocketHandler(final double idDouble) {
    final int id = (int) idDouble;

    WebSocketModule webSocketModule = getWebSocketModule("addWebSocketHandler");

    if (webSocketModule != null) {
      webSocketModule.setContentHandler(id, mWebSocketContentHandler);
    }
  }
```
That code adds the same singleton (`WebSocketModule`)'s reference as a handler for a given socket connection ID.
Because the handler will never change, it's useless to add references like that in our C++ code.
What we do in `addWebSocketHandler` is simpler:
```C++
       [contentHandler = m_contentHandler](dynamic args) {
         auto id = jsArgAsInt(args, 0);

         contentHandler->Register(id);
       }
```
We only keep track of the connection ID, for the object that will handle it is set for both `BlobModule` and the rest of the modules in the constructor by adding it to the property bag.

File: vnext/Shared/Networking/WinRTHttpResource.cpp
Comment (JunielKatarn) [reply vnext/Shared/Networking/WinRTHttpResource.cpp:198]: As discussed offline, not implementing custom URIs on this PR.

**Comment by JunielKatarn** [comment]
Ready to merge.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:133]: FileReader module has been implemented and tested (See `BobTest.js`).

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp]: URI handlers have been postponed, left out of this PR.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp]: Data is immutable and only exposed via `winrt::array_view<uint8_t>(std::vector<uint8_t>::data())`.

File: vnext/src/IntegrationTests/WebSocketBlobTest.js
Comment (mganandraj) [inline vnext/src/IntegrationTests/WebSocketBlobTest.js:114]: You can use FileReader to verify the message ..

File: vnext/Shared/OInstance.cpp
Comment (mganandraj) [reply vnext/Shared/OInstance.cpp:550]: Ah .. it looks the copy is shallow by default .. makes sense ..

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:53]: >extensionToMime = { [](http://example.com/codeflow?start=54&length=19)

This is unused ?

File: vnext/Shared/Modules/NetworkingModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/NetworkingModule.cpp:264]: >|| responseType == "blob" [](http://example.com/codeflow?start=60&length=25)

I think you you've added the blob support to HttpModule, not to NetworkingModule ? (I may be wrong)

File: vnext/Shared/Modules/FileReaderModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/FileReaderModule.cpp:109]: >ReactApplicationContext [](http://example.com/codeflow?start=57&length=23)

This is an Android entity

File: vnext/Shared/Modules/FileReaderModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/FileReaderModule.cpp:127]: Have a bug for this ..

File: vnext/Shared/Modules/BlobModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/BlobModule.cpp:308]: >return {}; [](http://example.com/codeflow?start=6&length=10)

Should we just return till the end here ?

File: vnext/Shared/Modules/WebSocketModule.cpp
Comment (mganandraj) [inline vnext/Shared/Modules/WebSocketModule.cpp:161]: >m_sharedState->InspectableProps [](http://example.com/codeflow?start=34&length=31)

WebSocketModuleTest create this module with "null" property bag

Comment (mganandraj) [reply vnext/Shared/Modules/WebSocketModule.cpp:161]: Both this and Networking module should handle the scenario where the blobs are not available .. or the property bag is set to null ..

File: vnext/src/IntegrationTests/WebSocketBlobTest.js
Comment (JunielKatarn) [reply vnext/src/IntegrationTests/WebSocketBlobTest.js:114]: Good idea.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:53]: Not now, but it will be needed when we implement custom URIs.
Removing for now.

File: vnext/Shared/Modules/NetworkingModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/NetworkingModule.cpp:264]: That is correct, but `NetworkingModule` can still be used by setting a specific runtime option.
It won't implement Blob functionality, but this avoids an unnecessary assert.

Context: This feature began development before switching to `HttpModule`.
Also, `NetworkingModule` will be removed soon.

File: vnext/Shared/Modules/FileReaderModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/FileReaderModule.cpp:127]: #9982 filed.

File: vnext/Shared/Modules/BlobModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/BlobModule.cpp:308]: Sounds reasonable.

File: vnext/Shared/Modules/WebSocketModule.cpp
Comment (JunielKatarn) [reply vnext/Shared/Modules/WebSocketModule.cpp:161]: The IInspectable arg should be guaranteed by InstanceImple (OInstance.cpp).

**Review by mganandraj — APPROVED** [review]
:shipit:

**Review by dannyvv — APPROVED** [review]

File: vnext/Shared/Networking/WinRTHttpResource.cpp
Comment (asklar) [inline vnext/Shared/Networking/WinRTHttpResource.cpp:71]: bug, `||`

**Pull request merged**
